src/mym.cpp

```diff
@@ -1800,3 +1808,83 @@ mxArray* deserialize(const char* rpSerial, const size_t rlength) {
         mxFree(p_cmp);
     return p_res;
 }
+char *hex2char(char *original_val, const size_t char_length) {
+    const uint8_t *pnt = (uint8_t *)original_val;
+    uint8_t *result_pnt = new uint8_t[char_length*4];
+    unsigned int offset = 0;
+    for( unsigned int a = 0; a < char_length; ++a )
+    {
+        if     (pnt[a]<=0x7F) {
+            result_pnt[a+offset] = pnt[a];
+        }
+        else if(pnt[a]<=0x7FF) {
+            result_pnt[a+offset] = ((pnt[a]>>6) + 0xC0);
+            result_pnt[a+offset+1] = ((pnt[a] & 0x3F) + 0x80);
+            offset += 1;
+        }
+        else if(0xD800<=pnt[a] && pnt[a]<=0xDFFF) {
+            mexErrMsgIdAndTxt("DataJoint:Deserialization:UTF8",
+                "Invalid block of UTF8 detected.");
+        } //invalid block of utf8
+        else if(pnt[a]<=0xFFFF) {
+            result_pnt[a+offset] = ((pnt[a]>>12) + 0xE0);
+            result_pnt[a+offset+1] = (((pnt[a]>>6) & 0x3F) + 0x80);
+            result_pnt[a+offset+2] = ((pnt[a] & 0x3F) + 0x80);
+            offset += 2;
+        }
+        else if(pnt[a]<=0x10FFFF) {
+            result_pnt[a+offset] = ((pnt[a]>>18) + 0xF0);
+            result_pnt[a+offset+1] = (((pnt[a]>>12) & 0x3F) + 0x80);
+            result_pnt[a+offset+2] = (((pnt[a]>>6) & 0x3F) + 0x80);
+            result_pnt[a+offset+3] = ((pnt[a] & 0x3F) + 0x80);
+            offset += 3;
+        }
+    }
+    result_pnt[char_length+offset]= 0x00;
+    return (char *)result_pnt;
+}
+char *char2hex(char *original_val, const size_t vlength, const size_t char_length) {
+    unsigned int idx = 0;
+    unsigned int curr_length = 0;
+    unsigned char u0,u1,u2,u3;
+    uint8_t *result_pnt = new uint8_t[char_length];
+    for(unsigned int a = 0; a < vlength;)
+    {
+        curr_length = 1;
+        if((original_val[a] & 0xF8) == 0xF0) curr_length = 4;
+        else if((original_val[a] & 0xF0) == 0xE0) curr_length = 3;
+        else if((original_val[a] & 0xE0) == 0xC0) curr_length = 2;
+        if((a + curr_length) > vlength) curr_length = 1;
+
+        u0 = original_val[a]; 
+        if (u0 >= 0x00 && u0 <= 0x7F) {
+            result_pnt[idx] = u0;
+            idx++;
+        }
+
+        u1 = original_val[a+1]; 
+        if (u0 >= 0xC0 && u0 <= 0xDF) {
+            result_pnt[idx] = (((u0-0xC0)<<6) + (u1-0x80));
+            idx++;
+        }
+        if (original_val[a] == 0xED && (original_val[a+1] & 0xA0) == 0xA0) {
+            mexErrMsgIdAndTxt("DataJoint:Serialization:UTF8",
+                "Invalid block of UTF8 detected.");
+        } //code points, 0xd800 to 0xdfff
+
+        u2 = original_val[a+2]; 

```

Comment:
a > &original_val[vlength] ...

this algorithm / loop structure needs to be adjusted to stay w/i array boundaries

Reply:
I see your point here. I can improve the algorithm to prevent such a scenario.

Reply:
curr_length is set based on the user data rather than the memory size and therefore the arrays can be violated  - take for example:

```
char original_val[] = { 0xf0, 0x00 };
````

might not happen in the nominal code path, but should be robust against this problem

Reply:
Ah, I understand now. To illustrate it with a realistic example, we could use the representation of the string `J`. The codepoint for this and the hex bytes are actually equivalent since it is within ASCII. Thus would be:
``` cpp
char original_val[] = { 0x4A, 0x00 };
size_t vlength = 1;
size_t char_length = 1;
```
For this case, the logic would violate the bounds. Thanks man and good catch. I will make an update to include this.

Reply:
I have update the logic to address this now.